File: src/main/python/com/hazza/LeetCodeByPython/SegmentTwo/No75_Sort_Colors.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-07-07
# https://leetcode.com/problems/sort-colors/
from typing import List


class Solution:
    def sortColors1(self, nums: List[int]) -> None:
        n0, n1 = 0, 0

        for n2, n in enumerate(nums[:]):
            nums[n2] = 2
            if n < 2:
                nums[n1] = 1
                n1 += 1
            if n == 0:
                nums[n0] = 0
                n0 += 1

    def sortColors2(self, nums: List[int]) -> None:
        n0, i, n2 = 0, 0, len(nums) - 1

        while i <= n2:
            if nums[i] == 0 and i != n0:
                nums[i], nums[n0] = nums[n0], nums[i]
                n0 += 1
            elif nums[i] == 2 and i != n2:
                nums[i], nums[n2] = nums[n2], nums[i]
                n2 -= 1
            else:
                i += 1


if __name__ == '__main__':
    solution = Solution()

    nums = [2, 0, 2, 1, 1, 0]
    solution.sortColors1(nums)
    assert nums == [0, 0, 1, 1, 2, 2]

    nums = [2, 0, 2, 1, 1, 0]
    solution.sortColors2(nums)
    assert nums == [0, 0, 1, 1, 2, 2]

File: src/main/java/com/hazza/leetcode/SegmentTwo/No75SortColors.java
package com.hazza.leetcode.SegmentTwo;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/sort-colors/discuss/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-08-28
 * Time: 9:43 AM
 */
public class No75SortColors {
    // solution 1
    public void sortColors(int[] nums) {
        int n0 = -1, n1 = -1, n2 = -1;
        int len = nums.length;
        for (int i = 0; i < len; i++) {
            if (nums[i] == 0) {
                nums[++n2] = 2;
                nums[++n1] = 1;
                nums[++n0] = 0;
            } else if (nums[i] == 1) {
                nums[++n2] = 2;
                nums[++n1] = 1;
            } else if (nums[i] == 2) {
                nums[++n2] = 2;
            }
        }
    }

    // solution 2
    public void sortColors1(int[] nums) {
        int num0 = 0, num2 = nums.length - 1;
        for (int i = 0; i <= num2; i++) {
            if (nums[i] == 0 && i != num0) swap(nums, i--, num0++);
            else if (nums[i] == 2 && i != num2) swap(nums, i--, num2--);
        }
    }

    private void swap(int[] nums, int i, int j) {
        if (nums[i] != nums[j]) {
            nums[i] ^= nums[j];
            nums[j] ^= nums[i];
            nums[i] ^= nums[j];
        }
    }
}
